### PR TITLE
[skillshare:class] Add support for skillshare.com classes

### PR DESCRIPTION
**Pull request by shaileshaanand**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/ytdl-org/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [x] New extractor
- [ ] New feature

---

### Add support for Skillshare.com classes

Add support for skillshare.com urls such as [This one](https://www.skillshare.com/classes/SEO-Today-Strategies-to-Earn-Trust-Rank-High-and-Stand-Out/423483018) using api calls reversed from the desktop website.

 - can only downlod entire classes, not single lessons as there are no separate urls for separate lessons
 - The website has Recaptcha on the login page, so only cookie authentication is supported

### CODE REVIEW

**Comment by dstftw** [comment]
1. Read coding conventions.
2. Extraction must be implemented in terms of brightcove extractor.
3. You must provide account credentials.

**Comment by shaileshaanand** [comment]
L

>     3. You must provide account credentials.
Login via credentials is not possible as there is Recaptcha on login page.
Do you mean credentials for testing?

**Comment by dstftw** [comment]
>Do you mean credentials for testing?

Exactly.

**Comment by shaileshaanand** [comment]
email: xafahe1300@topmail.ws
pass: D7Gp5TcDjUfkM2

**Comment by shaileshaanand** [comment]
>     1. Read coding conventions.
> 
>     2. Extraction must be implemented in terms of brightcove extractor.
> 
>     3. You must provide account credentials.

I have made the required changes, please verify.

**Comment by dstftw** [comment]
>Read coding conventions.

**Comment by shaileshaanand** [comment]
>     1. Read coding conventions.
Could you please point where I have violated coding conventions?
 - I have removed the unnecessary variables that were used only once.
 - I am using get() instead of try_get() as the data is not optional.
 - Should I add more fallbacks for title, description etc.?

**Comment by shaileshaanand** [comment]
Made more changes, please verify.

File: youtube_dl/extractor/skillshare.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    IE_NAME = 'skillshare:class'
+    _VALID_URL = r'https?://(?:www\.)?skillshare\.com/classes/[^/]+/(?P<id>[0-9]+)'
+    _TEST = {
+        'url': 'https://www.skillshare.com/classes/SEO-Today-Strategies-to-Earn-Trust-Rank-High-and-Stand-Out/423483018',
+        'only_matching': True,
+    }
+
+    def _real_extract(self, url):
+        class_id = self._match_id(url)
+        class_json_data = self._parse_json(self._search_regex(r'(?P<json_line>{"userData":{.+});\n', self._download_webpage(url, class_id), 'class_json_data'), class_id)
```
Comment (dstftw) [inline youtube_dl/extractor/skillshare.py]: Breaks once `userData` is not first.

Comment (dstftw) [inline youtube_dl/extractor/skillshare.py]: Carry long lines. Bother to finally read coding conventions.

File: youtube_dl/extractor/skillshare.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+        account_id = str(class_json_data['pageData']['videoPlayerData']['brightcoveAccountId'])
```
Comment (dstftw) [inline youtube_dl/extractor/skillshare.py]: No `str`.

File: youtube_dl/extractor/skillshare.py [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+        lessons = class_json_data['pageData']['videoPlayerData']['units'][0]['sessions']
+        entries = []
+        for lesson in lessons:
+            lesson_id = str(lesson['id'])
```
Comment (dstftw) [inline youtube_dl/extractor/skillshare.py]: Breaks.

File: youtube_dl/extractor/skillshare.py [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+            lesson_info_api_response = self._download_json("https://www.skillshare.com/sessions/{}/video".format(lesson_id), lesson_id)
```
Comment (dstftw) [inline youtube_dl/extractor/skillshare.py]: No `{}`.

File: youtube_dl/extractor/skillshare.py [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+            if 'video_hashed_id' not in lesson_info_api_response:
+                break
+            video_hashed_id = lesson_info_api_response['video_hashed_id'][3:]
```
Comment (dstftw) [inline youtube_dl/extractor/skillshare.py]: Breaks.

File: youtube_dl/extractor/skillshare.py [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+            entry = {
+                '_type': 'url_transparent',
+                'id': video_hashed_id,
```
Comment (dstftw) [inline youtube_dl/extractor/skillshare.py]: Useless.

File: youtube_dl/extractor/skillshare.py [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+                'title': lesson['title'],
```
Comment (dstftw) [inline youtube_dl/extractor/skillshare.py]: Breaks.

File: youtube_dl/extractor/skillshare.py [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+                'ie_key': BrightcoveNewIE.ie_key(),
+                'url': 'https://players.brightcove.net/{}/default_default/index.html?videoId={}'.format(account_id, video_hashed_id),
```
Comment (dstftw) [inline youtube_dl/extractor/skillshare.py]: Same.

File: youtube_dl/extractor/skillshare.py [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+            }
+            entries.append(entry)
+        return self.playlist_result(entries, class_id, class_json_data['pageData']['headerData']['title'], class_json_data.get("pageData").get('sectionData').get('description'))
```
Comment (dstftw) [inline youtube_dl/extractor/skillshare.py]: Playlist title is optional, description breaks.

File: youtube_dl/extractor/skillshare.py (hunk 6 above)
Comment (shaileshaanand) [reply youtube_dl/extractor/skillshare.py]: Is the id field useless?

Comment (dstftw) [reply youtube_dl/extractor/skillshare.py]: You are **delegating** to brightcove that provides the id.

File: youtube_dl/extractor/skillshare.py (hunk 5 above)
Comment (shaileshaanand) [reply youtube_dl/extractor/skillshare.py]: is ```[3:]``` the problem?

Comment (dstftw) [reply youtube_dl/extractor/skillshare.py]: Not only.

File: youtube_dl/extractor/skillshare.py (hunk 7 above)
Comment (shaileshaanand) [reply youtube_dl/extractor/skillshare.py]: Should I provide more fallbacks?
Should I use get() ?

Comment (dstftw) [reply youtube_dl/extractor/skillshare.py]: You must make it **not break** as you delegate.

Comment (shaileshaanand) [reply youtube_dl/extractor/skillshare.py]: Is the title field optional here?

File: youtube_dl/extractor/skillshare.py (hunk 5 above)
Comment (shaileshaanand) [reply youtube_dl/extractor/skillshare.py]: Should I use get() here?
Please tell me what is the other problem here, I am new to Python

**Comment by shaileshaanand** [comment]
Hopefully made all the requested changes. Please Verify

**Pull request closed without merging**